devito/mpi/distributed.py

```diff
@@ -432,22 +465,226 @@ def neighborhood(self):
 
         return ret
 
+    def _rebuild(self, shape=None, dimensions=None, comm=None):
+        return Distributor(shape or self.shape, dimensions or self.dimensions,
+                           comm or self.comm)
+
+
+class SubDistributor(DenseDistributor):
+
+    """
+    Decompose a subset of a domain over a set of MPI processes in a manner consistent
+    with the parent distributor.
+
+    Parameters
+    ----------
+    subdomain : SubDomain
+        The subdomain to be decomposed.
+
+    Notes
+    -----
+    This class is used internally by Devito for distributing
+    Functions defined on SubDomains. It stores reference to the
+    parent Distributor from which it is created.
+    """
+
+    def __init__(self, subdomain):
+        # Does not keep reference to the SubDomain since SubDomain will point to
+        # this SubDistributor and Distributor does not point to the Grid
+        super().__init__(subdomain.shape, subdomain.dimensions)
+
+        self._subdomain_name = subdomain.name
+        self._dimension_map = frozendict({pd: sd for pd, sd
+                                          in zip(subdomain.grid.dimensions,
+                                                 subdomain.dimensions)})
+
+        self._parent = subdomain.grid.distributor
+        self._comm = self.parent.comm
+        self._topology = self.parent.topology
+
+        self.__decomposition_setup__()
+
+    def __decomposition_setup__(self):
+        """
+        Set up the decomposition, aligned with that of the parent Distributor.
+        """
+        decompositions = []
+        for dec, i in zip(self.parent._decomposition, self.subdomain_interval):
+            if i is None:
+                decompositions.append(dec)
+            else:
+                start, end = _interval_bounds(i)
+                decompositions.append([d[np.logical_and(d >= start, d <= end)]
+                                       for d in dec])
+
+        self._decomposition = [Decomposition(d, c)
+                               for d, c in zip(decompositions, self.mycoords)]
+
+    @property
+    def parent(self):
+        """The parent distributor of this SubDistributor."""
+        return self._parent
+
+    p = parent  # Shortcut for convenience
```

Comment:
This seems like a bad idea
```suggestion
```

Reply:
Yeah, should probably be a property

Reply:
Does it need to be a single letter? I think parent is a lot more verbose/informative